### PR TITLE
Support multiple frustums

### PR DESCRIPTION
**Pull request by nithinp7**
Allows arbitrary list of view frustums to be passed in to the refinement engine.

### CODE REVIEW

**Comment by baothientran** [comment]
@nithinp7 can you separate the orthographic frustum commit in a separate PR?

**Review by javagl — COMMENTED** [review]
Two nitpicks inline, will inline further thoughts as comments, and still have to test it a bit

**Comment by javagl** [comment]
The current solution, from just reading it, seems to be a very pragmatic solution to the immediate goal.

---

Beyond that: The traversal is one of my pet peeves, with some elaborate discussion at https://github.com/CesiumGS/cesium-native/issues/126 . Some of the "shortcomings" that I mentioned there are not supposed to be affected by this PR. I think that some of the changes are not really improvements in that regard, but that might not be considered to be important, in view of the need to achieve a certain goal quickly (and that's what it does). 

But one thing that I wondered about: There have been methods like `bool isVisibleInFog(double distance, double fogDensity)` and `bool _meetsSse(const ViewState& viewState, const Tile& tile, double distance, bool culled)`, but also `addTileToLoadQueue` (I already talked about [the _meetsSse method before](https://github.com/CesiumGS/cesium-native/issues/126#issuecomment-774504271), and the load queues in general...). Now, these methods receive _vectors_ of `distances` and `frustums` (which actually are `viewStates`, but ... I'm not gonna talk about that one for now). When there already is a method 

    static bool isVisibleInFog(double distance, double fogDensity) {
        ...
    }

and then "that thing" has to be checked for multiple distances and densities, then I wonder whether there is a reason to pass in the vectors, when the given method already does this one, specific thing, and is documented and such. I'd then rather tend to *call* this method in another one:
```
static bool isVisibleInFog(
    const std::vector<ViewState>& frustums,
    const std::vector<double>& distances,
    const std::vector<FogDensityAtHeight>& fogDensityTable) {
    for (...) {
        double density = computeDensity(fogDensityTable, frustums[i]);
        if (!isVisibleInFog(distances[i], density) {
            return false;
        }
    }
    return true;
}
```
similar to how it was done for `isVisibleFromCamera`.  

The same applies to the `addTileToLoadQueue` method, but I see that it's a bit more difficult there: I already added a comment about some `priorityFor(tile, viewState, distance)`, because the statement: "In order to add a loading task to a queue, you need a distance" sounds odd, and more generally, I think there should not be three load queues (because that makes no sense), but ... I'll try to avoid further bragging about all the predictions that I made in https://github.com/CesiumGS/cesium-native/issues/126 coming true (e.g. "if we had a `TraversalState`, then this whole change could have been much, **much** simpler"), and will do some tests now instead.

---

**tl;dr**: If the tests don't bring up any issues, I have no further comments for now.

**Review by baothientran — COMMENTED** [review]
This looks really good @nithinp7! I only has some small comments below

File: Cesium3DTiles/src/Tileset.cpp [hunk 1]
```diff
@@ -1201,12 +1197,28 @@ static bool isVisibleFromCamera(
...
  */
-static bool isVisibleInFog(double distance, double fogDensity) {
-  if (fogDensity <= 0.0) {
-    return true;
+static bool isVisibleInFog(
+    const std::vector<ViewState>& frustums,
+    const std::vector<double>& distances,
+    const std::vector<FogDensityAtHeight>& fogDensityTable) {
+
+  for (size_t i = 0; i < frustums.size() && i < distances.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (baothientran) [reply Cesium3DTiles/src/Tileset.cpp:1205]: I agree this should be using assert as it's developer error at this point

**Comment by baothientran** [comment]
@nithinp7 can you also please add tests for multiple frustums?

File: Cesium3DTiles/src/Tileset.cpp [hunk 2]
```diff
@@ -1201,12 +1197,28 @@ static bool isVisibleFromCamera(
...
+    const ViewState& frustum = frustums[i];
+    const double& distance = distances[i];
+
+    double fogDensity = computeFogDensity(fogDensityTable, frustum);
```
Comment (baothientran) [inline Cesium3DTiles/src/Tileset.cpp:1209]: this may re-compute fog density for every tile even though fog-density only change between frustums

File: Cesium3DTiles/src/Tileset.cpp [hunk 3]
```diff
@@ -1201,12 +1197,28 @@ static bool isVisibleFromCamera(
...
+
+    double fogScalar = distance * fogDensity;
+    if (glm::exp(-(fogScalar * fogScalar)) > 0.0) {
+      return true;
+    }
   }
-  double fogScalar = distance * fogDensity;
-  return glm::exp(-(fogScalar * fogScalar)) > 0.0;
+
+  return false;
```
Comment (baothientran) [inline Cesium3DTiles/src/Tileset.cpp:1221]: should we fog-cull a tile if it is fog-culled by all the frustums? This looks like it culls a tile if the tile is fog-culled by one of the frustum

File: Cesium3DTiles/src/Tileset.cpp [hunk 4]
```diff
@@ -1868,24 +1911,34 @@ static bool anyRasterOverlaysNeedLoading(const Tile& tile) {
...
+
+      glm::dvec3 tileDirection = boundingVolumeCenter - frustum.getPosition();
+      double magnitude = glm::length(tileDirection);
+
+      if (magnitude >= CesiumUtility::Math::EPSILON5) {
+        tileDirection /= magnitude;
+        double loadPriority =
+            (1.0 - glm::dot(tileDirection, frustum.getDirection())) * distance;
+        if (loadPriority > maxLoadPriority) {
+          maxLoadPriority = loadPriority;
```
Comment (kring) [inline Cesium3DTiles/src/Tileset.cpp:1936]: I think you want the minimum load priority. Lower priority values load sooner.

File: Cesium3DTiles/src/Tileset.cpp [hunk 5]
```diff
@@ -1224,18 +1236,30 @@ Tileset::TraversalDetails Tileset::_visitTileIfNeeded(
...
-          boundingVolume,
-          this->_options.renderTilesUnderCamera)) {
+  const std::vector<ViewState>& frustums = frameState.frustums;
+
+  if (std::none_of(
+          frustums.begin(),
+          frustums.end(),
+          [boundingVolume,
+           renderTilesUnderCamera = this->_options.renderTilesUnderCamera](
+              const ViewState& frustum) {
```
[the remark this answers is not in the document]
Comment (kring) [reply Cesium3DTiles/src/Tileset.cpp:1257]: fwiw I quite like this functional-ish style.

File: Cesium3DTiles/src/Tileset.cpp (hunk 3 above)
Comment (kring) [reply Cesium3DTiles/src/Tileset.cpp:1221]: The logic looks correct to me. It returns true (visible) if its not in fog in any of the frustums, and returns false (culled) only if its in fog for all the frustums.

File: Cesium3DTiles/src/Tileset.cpp (hunk 4 above)
Comment (javagl) [reply Cesium3DTiles/src/Tileset.cpp:1936]: I also didn't spot that, although, as you might remember, I once criticized that "low `priority`" means "high priority". Things like this cause bugs, period.

File: Cesium3DTiles/src/Tileset.cpp (hunk 3 above)
Comment (baothientran) [reply Cesium3DTiles/src/Tileset.cpp:1222]: ah that's right! Sorry for the false alarm

File: Cesium3DTiles/src/Tileset.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (baothientran) [reply Cesium3DTiles/src/Tileset.cpp:1258]: I don't have strong opinion on this, so if everyone is good with it, then it's okay for me

[the remark this answers is not in the document]
Comment (javagl) [reply Cesium3DTiles/src/Tileset.cpp:1258]: Subjetively, I also like this functional style (and it has several advantages - there's a compelling reason for nearly all major programming languages having moved into that direction since 2005 - "The free lunch is over", roughly related to "internal iteration" vs "external iteration" and such...)

**But** ... I don't like large, anonymous lambdas. This specific block would IMHO be *much* better if it was just a line 
```
if (std::none_of(frustums.begin(), frustums.end(), tileIsInFrustum)) {
    ...
}
```
with a proper `auto tileIsInFrustum = ...`  lambda.

**Comment by nithinp7** [comment]
@baothientran @kring I added a few tests, I definitely need to write more tests for it but maybe that can be a separate PR for the sake of time? Otherwise, I think this works as expected. I'll note though that the frustum/fog culling is a bit awkward and maybe could be better optimized for cases where the different frustums are on different sides of the world. It would also be awkward to add a hardware occlusion query stage to this the way it is currently structured, so we may want to slightly refactor that code or at least create an issue for doing so.

File: Cesium3DTiles/src/Tileset.cpp [hunk 6]
```diff
@@ -1201,10 +1216,11 @@ static bool isVisibleFromCamera(
  * @param fogDensity The fog density
  * @return Whether the tile is visible in the fog
  */
-static bool isVisibleInFog(double distance, double fogDensity) {
+static bool isVisibleInFog(const double& distance, const double& fogDensity) {
```
Comment (kring) [inline Cesium3DTiles/src/Tileset.cpp:1219]: Why the change here? It's generally better to pass doubles by value.
